lib/network/webserver.js

```diff
@@ -0,0 +1,348 @@
+     * @return {WebServer} itself
+     */
+    onError(cb) {
+        assert.strictEqual(typeof cb, 'function',
+            'Callback must be a function');
+        this._cbOnError = cb;
+        return this;
+    }
+
+    /**
+     * Set the stop handler callback
+     *
+     * @param {function} cb - Callback()
+     * @return {WebServer} itself
+     */
+    onStop(cb) {
+        assert.strictEqual(typeof cb, 'function',
+            'Callback must be a function');
+        this._cbOnStop = cb;
+        return this;
+    }
+
+    /**
+     * Function called when a secure connection is etablished
+     *
+     * @param {tls.TlsSocket} sock - socket
+     * @return {undefined}
+     */
+    _onSecureConnection(sock) {
+        if (!sock.authorized) {
+            this._logger.error('rejected secure connection', {
+                address: sock.address(),
+                authorized: false,
+                error: sock.authorizationError,
+            });
+        }
+    }
+
+    /**
+     * function called when an error came from the client request
+     *
+     * @param {Error} err - Error
+     * @param {net.Socket|tls.TlsSocket} sock - Socket
+     * @return {undefined}
+     */
+    _onClientError(err, sock) {
+        this._logger.error('client error', {
+            method: 'arsenal.network.WebServer._onClientError',
+            error: err,
+            address: sock.address(),
+        });
+    }
+
+    /**
+     * Function to start the webserver
+     *
+     * @return {WebServer} itself
+     */
+    start() {
+        if (!this._server) {
+            if (this.isHttps) {
+                this._logger.info('starting webserver under https', {
+                    method: 'arsenal.network.WebServer.start',
+                    port: this._port,
+                });
+                this._https.agent = new https.Agent(this._https);
+                this._server = https.createServer(this._https,
+                    (req, res) => this._onRequest(req, res));
+            } else {
+                this._logger.info('starting webserver under http', {
+                    method: 'arsenal.network.WebServer.start',
+                    port: this._port,
+                });
+                this._server = http.createServer(
+                    (req, res) => this._onRequest(req, res));
+            }
+        }
+        this._server.on('error', err => this._onError(err));
+        this._server.on('secureConnection',
+            sock => this._onSecureConnection(sock));
+        this._server.on('tlsClientError', (err, sock) =>
+            this._onClientError(err, sock));
+        this._server.on('clientError', (err, sock) =>
+            this._onClientError(err, sock));
+        this._server.listen(this._port, () => this._onListening());
+        return this;
+    }
+
+    /**
+     * Function to stop the webserver
+     *
+     * @return {WebServer} itself
+     */
+    stop() {
+        if (this.server) {
```

Comment:
can you explicitly call the function like ```this.server()``` on each ref of ```this.server``` . 
Because i'm agree with @rahulreddy  it's a bit confusing .